File: android/app/src/main/java/com/redcarrottt/sc/Device.java
package com.redcarrottt.sc;

/* Copyright (c) 2017-2018. All rights reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import com.redcarrottt.testapp.Logger;

public abstract class Device {
    private final String kTag = "Device";

    // Main Functions
    @SuppressWarnings("SynchronizeOnNonFinalField")
    boolean holdAndTurnOn() {
        int refCount;
        synchronized (this.mRefCount) {
            this.mRefCount++;
            refCount = this.mRefCount;
        }
        if (refCount == 1) {
            this.setState(State.kTurningOn);
            boolean res = this.turnOnImpl();
            if (!res) {
                synchronized (this.mRefCount) {
                    this.mRefCount--;
                }
                this.setState(State.kOff);
            } else {
                this.setState(State.kOn);
            }
            Logger.VERB(kTag, this.mName + " is turned on");
            return res;
        } else {
            // Already turn on
            Logger.VERB(kTag, this.mName + " is already turned on");
            return true;
        }
    }

    @SuppressWarnings("SynchronizeOnNonFinalField")
    boolean releaseAndTurnOff() {
        int refCount;
        synchronized (this.mRefCount) {
            this.mRefCount--;
            refCount = this.mRefCount;
        }
        if (refCount == 0) {
            this.setState(State.kTurningOff);
            boolean res = this.turnOffImpl();
            if (!res) {
                synchronized (this.mRefCount) {
                    this.mRefCount++;
                }
                this.setState(State.kOn);
            } else {
                this.setState(State.kOff);
            }
            Logger.VERB(kTag, this.mName + " is turned off");
            return res;
        } else {
            // Not yet turn off
            Logger.VERB(kTag, this.mName + " is used by other components, so not turned off");
            return true;
        }
    }

    // Implemented by child classes
    protected abstract boolean turnOnImpl();

    protected abstract boolean turnOffImpl();

    // State
    class State {
        public static final int kOff = 0;
        public static final int kTurningOn = 1;
        public static final int kOn = 2;
        public static final int kTurningOff = 3;
    }

    @SuppressWarnings("SynchronizeOnNonFinalField")
    protected int getState() {
        int state;
        synchronized (this.mState) {
            state = this.mState;
        }
        return state;
    }

    @SuppressWarnings("SynchronizeOnNonFinalField")
    private void setState(int newState) {
        synchronized (this.mState) {
            this.mState = newState;
        }
    }

    public Device(String name) {
        this.mState = State.kOff;
        this.mName = name;
    }

    // Attributes
    private String mName;
    private Integer mRefCount;

    // State
    private Integer mState;
}

File: linux/src/Device.cpp
/* Copyright 2017-2018 All Rights Reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *  
 * [Contact]
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0(the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <ServerAdapter.h>

using namespace sc;

bool Device::hold_and_turn_on(void) {
  if(this->mRefCount.increase() == 1) {
    this->set_state(DeviceState::kTurningOn);

    bool res = this->turn_on_impl();

    if(!res) {
      this->mRefCount.decrease();
      this->set_state(DeviceState::kOff);
    } else {
      this->set_state(DeviceState::kOn);
    }
    return res;
  } else {
    // Already turn on
    return true;
  }
}

bool Device::release_and_turn_off(void) {
  if(this->mRefCount.decrease() == 0) {
    this->set_state(DeviceState::kTurningOff);

    bool res = this->turn_off_impl();

    if(!res) {
      LOG_DEBUG("%s: Failed to turn off", this->get_name());
      this->mRefCount.increase();
      this->set_state(DeviceState::kOn);
    } else {
      LOG_DEBUG("%s: Successfully turned off", this->get_name());
      this->set_state(DeviceState::kOff);
    }
    return res;
  } else {
    // Not yet turn off
    LOG_DEBUG("%s: Not yet turn off", this->get_name());
    return true;
  }
}
